posts/Yoann/2017-11-27-introducing-hibernate-search-jsr352-mass-indexing-job.adoc

```diff
@@ -0,0 +1,163 @@
+
+In a Java SE application, you can use the reference implementation for JSR-352,
+https://github.com/WASdev/standards.jsr352.jbatch#readme[JBatch], in Java SE mode.
+You will need to:
+
+* add a dependency to the following artifacts in your project:
+** `org.hibernate:hibernate-search-jsr352-core:5.9.0.Beta1`
+** `com.ibm.jbatch:com.ibm.jbatch-runtime:1.0`
+** `javax.batch:javax.batch-api:1.0`
+** `javax.inject:javax.inject:1`
+** `org.apache.derby:derby:10.13.1.1`
+* add https://github.com/hibernate/hibernate-search/tree/master/integrationtest/jsr352/src/test/resources/META-INF/services[configuration files]
+  to you classpath under `META-INF/services`,
+* retrieve the `JobOperator` in your application by calling `javax.batch.runtime.BatchRuntime.getJobOperator()`
+(be sure to store the result, as each call may create a new operator).
+* make sure to initialize Hibernate ORM before you try to launch the mass indexing job,
+  and to only shut it down after the last mass indexing job stopped.
+
+As for other JSR-352 runtimes, the mass indexer may or may not work.
+We are in the process of testing at least one other JEE application server and Spring Batch:
+see <<limitations-and-next>> below.
+
+=== Usage
+
+All the examples below assume the variable `jobOperator` contains
+the job operator retrieved as instructed above.
+
+To launch a mass indexing job:
+
+```
+java.util.Properties parameters = org.hibernate.search.jsr352.massindexing.MassIndexingJob.parameters()
+		.forEntities( MyFirstEntity.class, MySecondEntity.class )
+		.build();
+long executionId = jobOperator.start(
+		org.hibernate.search.jsr352.massindexing.MassIndexingJob.NAME,
+		parameters
+		);
+```
+
+You can then query the execution status through this call:
+
+```
+javax.batch.runtime.JobExecution jobExecution = jobOperator.getJobExecution( executionId );
+```
+
+You can suspend the job this way:
+
+```
+long executionId2 = jobOperator.stop( executionId );
+```
+
+To resume the job after suspension or upon failure, you can request a restart this way:
+
+```
+long executionId2 = jobOperator.restart( executionId, parameters );
+```
+
+[TIP]
+====
+According to the JSR-352 specification, you must provide the parameters again when restarting a job.
+
+However, https://github.com/jberet/jsr352#readme[JBeret], the JSR-352 implementation included in Wildfly,
+allows you to pass `null` instead, and will interpret it as
+"the parameters I used when initially starting the job".
+====
+
+The builder you get from calling `MassIndexingJob.parameters()` offers a few settings.
+Here are some of the most useful:
+
+* `.purgeAllOnStart(boolean)`: specify whether the existing index should be purged at the beginning of the job.
+* `.restrictedBy(Criterion)`: use a `Criterion` to only index a subset of entities.
+* `.restrictedBy(String)`: use HQL to only index a subset of entities.
+  Note that when using HQL, the job switches to a degraded mode that suffers from several limitations,
```

Comment:
That sounds mysterious, what is it about?

Reply:
Mainly no partitioning (because we don't use a list of IDs to drive the indexing, but the result of a query returning entities). So HQL indexing will use a single thread. We might try to change that, rely on limit/offset to run the query in multiple threads, but then there will be a risk of indexing some entities twice or skipping some entities. Either way, not a good idea to use HQL when you could use something else.